File: backend/app/tokens.py
"""The one token-counting rule, shared by ingestion and the chat budget (T-310).

Promoted out of `app.ingestion.chunker`, where it lived alone while `document_chunks.
token_count` was its only consumer. T-310 gives it a second and a **normative** one: the
NFR-CAP-01 budget FR-STA-04 enforces and FR-ANL-03 displays. Two subsystems applying the
same rule from two definitions is how the number a user is shown stops matching the number
that blocked them, so there is exactly one here — the `split_answer_segments` argument
(R-48(4)) applied to arithmetic.

**It is an estimate, and that is a recorded deviation** — FR-ANL-03 says "production shall
use real token usage", which R-30 has already reinterpreted once: the API reports a single
`prompt_tokens` covering the system prompt, the retrieved chunks, the history and the query,
while R-30 counts only the last two, so **no API-reported number can produce this meter**.
What FR-ANL-03 was contrasting against is the prototype's fixed `+220`/`+640` simulation,
and counting real message text satisfies that intent.

Three things make the imprecision safe rather than merely tolerable:

* R-30(2) makes 10.4K a **product** budget, not a model limit — the OpenAI models carry far
  larger windows — so an error here can never turn into a provider 400. It moves where a
  chat stops, not whether it works.
* The rule has to run **in the browser too**: FR-STA-04 blocks *before* submission, and the
  GUI cannot ask the server to count every keystroke. `ceil(len/4)` is one line of
  TypeScript; `tiktoken` is a WASM bundle, and a mismatch between the two sides would show
  as the composer refusing a message the server would have accepted.
* R-35(7) already refused `tiktoken` for the worker because it fetches its BPE vocabulary
  over the network on first use — a cold-start hazard that is worse, not better, on a
  request path.

Swapping in an exact tokenizer later is a change to this module and to the GUI's copy of the
rule, and nothing else — no stored value depends on it (`token_count` is not an FR-ING-03
fingerprint input, and the budget is derived per request, never persisted).
"""

from __future__ import annotations

import math

__all__ = ["CHARS_PER_TOKEN", "estimate_tokens"]

#: Characters per token. English prose against the GPT tokenizers sits near four; the error
#: is a few percent on chat text and larger on code or dense punctuation. # TBD(§8.4)
CHARS_PER_TOKEN = 4.0


def estimate_tokens(text: str) -> int:
    """Approximate the token count of `text`.

    Rounds **up**, and is applied **per message** rather than to a concatenated transcript:
    `sum(ceil(len_i / 4))` and `ceil(sum(len_i) / 4)` differ, and the per-message form is the
    one the GUI can reproduce for a single composer box without knowing the whole chat.

    T-205 must **not** size embedding batches from this — an underestimate becomes a 400 from
    the API. Batch by characters or by request count and let the API be the authority.
    """
    return math.ceil(len(text) / CHARS_PER_TOKEN)

File: frontend/src/tokens.ts
/**
 * The browser half of the one token-counting rule — `backend/app/tokens.py`, in TypeScript.
 *
 * R-51(2) requires this to exist and requires it to be *this* rule rather than a better one:
 * FR-STA-04 blocks **before** submission, so the GUI cannot ask the server to count every
 * keystroke, and a mismatch between the two sides shows up as the composer refusing a message
 * the server would have accepted (or, worse, accepting one it refuses — an unpredicted `409`,
 * which is the shape OI-31 already records as unspecified).
 *
 * So this is a deliberate second copy of a rule, and the only defensible kind: `ceil(len / 4)`
 * is one line, has no dependency, and cannot drift *silently* because `tokens.test.ts` reads
 * `backend/app/tokens.py` off disk and fails when the two constants part company. An exact
 * tokenizer would be a WASM bundle here and a network fetch there (R-35(7)), and the two
 * implementations would then disagree in ways no test could pin.
 *
 * **Estimate, by ruling.** R-30(2) makes 10.4K a *product* budget an order of magnitude below
 * the model window, so an error here moves where a chat stops — never whether a request works.
 */

/**
 * Characters per token, and it must equal `CHARS_PER_TOKEN` in `backend/app/tokens.py`.
 * Changing it moves the number in front of the user, so it changes in both places in one
 * commit (R-51(2)). `# TBD(§8.4)` on the backend side; the same TBD, not a second one.
 */
export const CHARS_PER_TOKEN = 4.0;

/**
 * Approximate the token count of `text`.
 *
 * Rounds **up**, and is applied **per message** — `sum(ceil(len_i / 4))` and
 * `ceil(sum(len_i) / 4)` are different numbers, and only the per-message form is one this side
 * can reproduce for a lone composer box without knowing the whole transcript.
 */
export function estimateTokens(text: string): number {
  return Math.ceil(text.length / CHARS_PER_TOKEN);
}

/** What `check_submission` returns, minus the parts only the server needs. */
export interface SubmissionCheck {
  /** False when FR-STA-04 must block. */
  allowed: boolean;
  /** `used + query + reserve` — what the turn would cost once answered. */
  projectedTokens: number;
  queryTokens: number;
}

/**
 * FR-STA-04 for one attempted message. Pure, and **identical to `check_submission`**.
 *
 * The projection is `used + query + reserve`, not `used + query`: the reply is part of the
 * conversation the instant it lands, so without the reserve a chat just under the limit
 * accepts a query, answers it, and is over budget with no block ever having fired (R-51(3)).
 *
 * **The comparison is `<=`, not `<`** — a projection landing exactly on the limit is inside
 * it. Ties are permissive, matching `check_submission` verbatim; getting this backwards would
 * make the composer refuse the one message the server accepts.
 *
 * `reserve` is the server's `CONTEXT_ANSWER_RESERVE_TOKENS`, floored at `LLM_MAX_OUTPUT_TOKENS`
 * by a cross-group validator. It is a **parameter rather than a constant here on purpose**: it
 * is an operator setting, so hard-coding 1500 would silently desynchronise this check from the
 * server the moment the answer ceiling is raised.
 */
export function projectSubmission(
  query: string,
  { used, limit, reserve }: { used: number; limit: number; reserve: number },
): SubmissionCheck {
  const queryTokens = estimateTokens(query);
  const projectedTokens = used + queryTokens + reserve;
  return { allowed: projectedTokens <= limit, projectedTokens, queryTokens };
}

/**
 * Whether the conversation is **frozen** in R-67(1)'s sense: no message of any length fits, so
 * the chat is at its terminal state rather than merely refusing this draft.
 *
 * Distinct from `!projectSubmission(...).allowed`, which is per-draft and clears when the user
 * shortens what they typed. FR-STA-04 says "block until content is reduced"; R-67(1) is the
 * ruling that the *history* is not reducible, so once even an empty query overflows there is
 * nothing left to reduce and the GUI must say so instead of inviting an edit.
 */
export function isFrozen(usage: { used: number; limit: number; reserve: number }): boolean {
  return !projectSubmission('', usage).allowed;
}
